File: TP2/src/couleurs.c
/* Fichier: phrases.c
 * Recherche de phrases dans un fichier
 * Auteur: Pierre MOLY / Maxence LERDA
 * Exercice 2.7
 * https://koor.fr/C/cstdlib/rand.wp
 * https://stackoverflow.com/questions/12344814/how-to-print-unsigned-char-as-2-digit-hex-value-in-c
 */

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

// Structure pour représenter une couleur
struct Couleur {
    int r;
    int g;
    int b;
    int a;
};

// Structure pour compter les occurrences des couleurs
struct Couleur_Count {
    struct Couleur couleur;
    int count;
};

// Fonction pour comparer deux couleurs
int compareColor(struct Couleur c1, struct Couleur c2) {
    // Comparaison des composantes RGB et alpha
    return (c1.r == c2.r && c1.g == c2.g && c1.b == c2.b && c1.a == c2.a);
}

int main() {

    srand(time(NULL));  // Initialisation du générateur de nombres aléatoires

    struct Couleur couleurs[100]; // Tableau de 100 couleurs
    struct Couleur_Count couleurs_count[100]; // Tableau de 100 couleurs et leurs occurrences
    int distinct_count = 0;  // Compteur de couleurs distinctes

    // Génération de 100 couleurs aléatoires
    for (int i = 0; i < 100; i++) {
        if (i % 10 == 0) {
            // Creation de couleurs égale pour tester le comptage
            couleurs[i].r = 12;
            couleurs[i].g = 12;
            couleurs[i].b = 12;
            couleurs[i].a = 255;
        } else {
            // Génération de couleurs aléatoires
            couleurs[i].r = rand() % 256;
            couleurs[i].g = rand() % 256;
            couleurs[i].b = rand() % 256;
            couleurs[i].a = 255;
        }
    }

    // Comptage des occurrences des couleurs distinctes
    for (int j = 0; j < 100; j++) {
        // Initialisation de la variable found
        int found = 0;

        // Chercher si la couleur existe déjà dans le tableau des couleurs distinctes
        for (int k = 0; k < distinct_count; k++) {
            if (compareColor(couleurs[j], couleurs_count[k].couleur)) {
                // Incrémenter le compteur de la couleur
                couleurs_count[k].count++;
                found = 1;
                // Sortir de la boucle
                break;
            }
        }

        // Si la couleur n'a pas été trouvée, l'ajouter au tableau des couleurs distinctes
        if (found == 0) {
            couleurs_count[distinct_count].couleur = couleurs[j];
            couleurs_count[distinct_count].count = 1;
            // Incrémenter le compteur de couleurs distinctes
            distinct_count++;
        }
    }

    // Affichage des couleurs distinctes et leurs occurrences
    for (int l = 0; l < distinct_count; l++) {
        // Affichage de la couleur et de son compteur en hexadécimal
        printf("%02X 0x%02X 0x%02X 0x%02X : %d \n", 
           couleurs_count[l].couleur.a, couleurs_count[l].couleur.r, 
           couleurs_count[l].couleur.g, couleurs_count[l].couleur.b, 
           couleurs_count[l].count);
    }

    return 0;
}

File: TP2/src/couleurs.py
# Fichier: couleurs.py
# Gestion des couleurs et comptage des occurrences
# Auteur: Pierre MOLY / Maxence LERDA
# Exercice 2.7

# Références:
# http://www.science-du-numerique.fr/comment-convertir-un-nombre-decimal-en-hexadecimal-en-python#:~:text=Pour%20convertir%20un%20nombre%20d%C3%A9cimal,repr%C3%A9sentant%20le%20nombre%20en%20hexad%C3%A9cimal.

from random import randint

class Couleur:
    # Classe pour stocker les informations d'une couleur
    def __init__(self, r: str, g: str, b: str, a:str) -> None:
        self.r = r
        self.g = g
        self.b = b
        self.a = a

def main():
    # Création d'une liste de 99 couleurs aléatoires
    color_list = []

    for i in range(100):

        if (i > 30 and i < 35):
            r = hex(212)
            g = hex(42)
            b = hex(78)
        else:
            # Génération de valeurs aléatoires pour les composantes R, G et B
            r = hex(randint(0, 255))
            g = hex(randint(0, 255))
            b = hex(randint(0, 255))
            # Création d'une instance de la classe Couleur
        couleur = Couleur(r, g, b, 'ff')
        # Ajout de la couleur à la liste
        color_list.append(couleur)

    dict_colors = {}
    # Comptage des occurrences de chaque couleur
    for color in color_list:
    # Si la couleur n'est pas dans le dictionnaire, on l'ajoute avec une occurrence de 1
        if dict_colors.get(f'{color.a} {color.r} {color.g} {color.b}') == None:
            dict_colors[f'{color.a} {color.r} {color.g} {color.b}'] = 1
        else:
            # Sinon, on incrémente l'occurrence de 1
            dict_colors[f'{color.a} {color.r} {color.g} {color.b}'] += 1
    
    for i in dict_colors:
        # Affichage des couleurs et de leur occurrence
        print(f"{i} : {dict_colors[i]}")


if __name__ == "__main__":
    main()
        

